src/aes.rs

```diff
@@ -0,0 +1,276 @@
+    let (aes_dek, aes_edek) = generate_aes_edek(rng, kek, id)?;
+    Ok((
+        aes_dek,
+        create_signed_header(
+            icl_header_v4::v4document_header::EdekWrapper {
+                edek: Some(Edek::Aes256GcmEdek(aes_edek)),
+                ..Default::default()
+            },
+            aes_dek,
+        ),
+    ))
+}
+
+/// Decrypt the aes edek. Does not verify signature of the header or check that the id is appropriate.
+/// You must do that as a separate step.
+pub fn decrypt_aes_edek(
+    kek: &EncryptionKey,
+    aes_edek: &icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek,
+) -> Result<EncryptionKey> {
+    let iv = aes_edek.iv.as_ref().try_into().map_err(|_| {
+        Error::DecryptError("IV from the edek was not the correct length.".to_string())
+    })?;
+    aes_decrypt(kek, iv, &aes_edek.ciphertext, &[])
+        .and_then(|dek_bytes| {
+            dek_bytes
+                .try_into()
+                .map_err(|_| Error::DecryptError("iv was not of the correct size".to_string()))
+        })
+        .map(|dek_bytes| EncryptionKey(dek_bytes))
+}
+
+/// Decrypt a V4 detatched document. The document should have the expected header
+pub fn decrypt_detatched_document(
+    key: &EncryptionKey,
+    payload: EncryptedPayload,
+) -> Result<PlaintextDocument> {
+    let payload_len = payload.0.len();
+    if payload_len < DETATCHED_HEADER_LEN + IV_LEN {
+        Err(Error::EdocTooShort(payload_len))
+    } else {
+        let header = payload.0.slice(0..DETATCHED_HEADER_LEN);
+        if header != [&[V0], &MAGIC[..]].concat() {
+            Err(Error::NoIronCoreMagic)
+        } else {
+            let iv = payload
+                .0
+                .slice(DETATCHED_HEADER_LEN..(DETATCHED_HEADER_LEN + IV_LEN))
+                .as_ref()
+                .try_into()
+                .expect("IV conversion will always have 12 bytes.");
+            let ciphertext = payload.0.slice(DETATCHED_HEADER_LEN + IV_LEN..);
+            aes_decrypt(key, iv, &ciphertext[..], &[]).map(|v| PlaintextDocument(v))
+        }
+    }
+}
+
+/// Encrypt a document to be used as a detatched document. This means it will have a header of `0IRON` as the first
+/// 5 bytes.
+pub fn encrypt_detatched_document<R: RngCore + CryptoRng>(
+    rng: &mut R,
+    key: EncryptionKey,
+    document: PlaintextDocument,
+) -> Result<EncryptedPayload> {
+    let (iv, enc_data) = aes_encrypt(key, &document.0, &[], rng)?;
+    let payload = EncryptedPayload(
+        [&[V0], &MAGIC[..], &iv[..], &enc_data.0[..]]
+            .concat()
+            .into(),
+    );
+    Ok(payload)
+}
+pub(crate) fn aes_encrypt<R: RngCore + CryptoRng>(
+    key: EncryptionKey,
+    plaintext: &[u8],
+    associated_data: &[u8],
+    rng: &mut R,
+) -> Result<([u8; 12], EncryptedDocument)> {
+    let cipher =
+        Aes256Gcm::new_from_slice(&key.0).expect("Key length 32 is always valid for Aes256Gcm");
```

Comment:
Can we use this instead?
```suggestion
    let cipher = Aes256Gcm::new(&key.0.into());
```